src/passes/LegalizeJSInterface.cpp

```diff
@@ -241,38 +242,53 @@ struct LegalizeJSInterface : public Pass {
     return func->name;
   }
 
-  void ensureTempRet0(Module* module) {
+
+  void addTempRet0Helpers(Module* module) {
+    bool getterExists = module->getExportOrNull(GET_TEMP_RET_0) != nullptr;
+    bool setterExists = module->getExportOrNull(SET_TEMP_RET_0) != nullptr;
+
     if (!module->getGlobalOrNull(TEMP_RET_0)) {
+      if (getterExists)
+        Fatal() << "tempRet0 global not found but getTempRet0 exists";
+      if (setterExists)
+        Fatal() << "tempRet0 global not found but setTempRet0 exists";
       module->addGlobal(Builder::makeGlobal(
         TEMP_RET_0,
         i32,
         LiteralUtils::makeZero(i32, *module),
         Builder::Mutable
       ));
-      needTempRet0Helpers = true;
     }
-  }
 
-  void addTempRet0Helpers(Module* module) {
-    // We should also let JS access the tempRet0 global, which
-    // is necessary to send/receive 64-bit return values.
+    // We should also let JS access the tempRet0 global, which is necessary
+    // to (among other things) send/receive 64-bit return values.
     auto exportIt = [&](Function* func) {
+      // Remove any existing import
+      Function* existing = module->getFunctionOrNull(func->name);
```

Comment:
This doesn't remove an import - it removes a function with that name, whether it's an import or not. Maybe I'm missing something here, but it seems like this tramples any existing function with that name?

I think it would be safe if both the getter and setter already exist, as long as they appear as a pair. The problem is if there is just one.

Reply:
I've added an assert that its an import.   Since its an export we already existing early from this function. 

I thought that imported functions were just function in the new model?  Is there some way to explicitly remove a function export?

Reply:
Imported functions are just functions, correct. However, these two are not imports, I think? We define get/setTempRet0 in the wasm itself.

Reply:
They can be imports in the case of SIDE_MODULE I believe.    Some of the test cases import them from env.  I

Reply:
Ah yes, for the dynamic library case we import them. So I guess we need to handle the case they are either imports or are not, if they exist.

Reply:
Btw, by coincidence I noticed a bunch of other bugs in this file, for https://github.com/kripken/emscripten/issues/7304 (dynamic linking stuff). Some overlap with this issue actually. I can either wait until this is landed, or I could do it all in one PR, whichever you prefer.

Reply:
You are right, this is more complicated than I thought.

These functions are also improted by the wasm backend codegen if this change lands:
https://reviews.llvm.org/D53240

(which I think it should).

If you want to take a look at cleaning all this on that would be great.  I was having some more thoughts about it this morning.

Here is how I think it should go:

1.  We should never create these function as part of binaryen
2.  We should assume they exist and import them if they don't
3.  We rely on the embedder to provide them, or for them to be compiled in at build time (this is the RELOCATABLE vs non-RELOCATABLE case)

This way we can remove the codegen from binaryen.   In the RELOCATABLE case these will end up as imports.  In the none-RELOCATABLE case these will be defined in the the JS code.